### PR TITLE
GitLab integration

### PR DESCRIPTION
**Pull request by isharshun**
* Add GitLab plugin
* Add Changeset support for GitLab projects as first part of plugin

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link:

### Risks
- Level: Low

### CODE REVIEW

File: app/models/changeset.rb [hunk 1]
```diff
@@ -71,11 +72,20 @@ def find_comparison
...
+          end
       end
 
       Rails.cache.fetch(cache_key) do
-        GITHUB.compare(repo, previous_commit, commit)
+        if project.gitlab?
+          Presenters::GitlabChangeset.new(Gitlab.compare(repo, previous_commit, commit)).build
+        else
+          GITHUB.compare(repo, previous_commit, commit)
+        end
```
Comment (grosser) [inline app/models/changeset.rb]: ideally these 2 would be hooks, but good enough for now ...

File: plugins/gitlab/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+GitLab integration.
```
Comment (grosser) [inline plugins/gitlab/README.md]: Maybe say a bit more like "Allows deploying gitlab projects, viewing their diff, blah blah"

File: plugins/gitlab/app/models/presenters/gitlab_commit.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+module Presenters
```
Comment (grosser) [inline plugins/gitlab/app/models/presenters/gitlab_commit.rb]: would prefer `Gitlab::CommitPresenter`

**Comment by grosser** [comment]
this looks pretty clean, did not think that was possible without rewiring big parts of the app ...

got some UI screenshots too ?

**Comment by isharshun** [comment]
UI is the same as for GitHub projects.
image
image

Sample error message for the GitLab API:
image

File: .env.test [hunk 4]
```diff
@@ -27,3 +27,4 @@ DOCKER_REGISTRIES=docker-registry.example.com
 DEPRECATED_TOKEN_AUTH=1
 GCLOUD_GKE_CLUSTERS_FOLDER=/tmp
 CACHE_STORE=memory
+GITLAB_TOKEN=
```
Comment (grosser) [inline .env.test]: newline after :D

**Comment by grosser** [comment]
let me know when you are happy with it and I'll merge ❤️

**Comment by isharshun** [comment]
Thanks, I`m happy this to be merged to master!

**Comment by grosser** [comment]
thx so much ... already 2+ failed PRs for this :)

**Comment by grosser** [comment]
For more integrations we should have a dedicated remote_repository on the project and then plug in whichever we need, hooks are kinda the wrong tool for the job (then we can also support "no remote = local" and have fallback logic there)

**Pull request merged**